embedding/csharp/Tizen.Flutter.Embedding/Channels/MethodChannel.cs

```diff
@@ -0,0 +1,156 @@
+        {
+            InvokeMethod(new MethodCall(method, arguments));
+        }
+
+        /// <summary>
+        /// Invokes a method on this channel, expecting no result.
+        /// </summary>
+        /// <param name="call">A <see cref="MethodCall"/>.</param>
+        public void InvokeMethod(MethodCall call)
+        {
+            if (call == null)
+            {
+                throw new ArgumentNullException(nameof(call));
+            }
+            BinaryMessenger.Send(Name, Codec.EncodeMethodCall(call));
+        }
+
+        /// <summary>
+        /// Invokes a method on this channel, expecting a result.
+        /// </summary>
+        /// <remarks>
+        /// A <see cref="FlutterException"/> is thrown, if the invocation failed in the Flutter application.
+        /// </remarks>
+        /// <param name="method">The name string of the method.</param>
+        /// <param name="arguments">The arguments for the invocation, possibly null.</param>
+        /// <returns>A <see cref="Task"/> that completes with a result.</returns>
+        public Task<object> InvokeMethodAsync(string method, object arguments)
+        {
+            return InvokeMethodAsync(new MethodCall(method, arguments));
+        }
+
+        /// <summary>
+        /// Invokes a method on this channel, expecting a result.
+        /// </summary>
+        /// <remarks>
+        /// A <see cref="FlutterException"/> is thrown, if the invocation failed in the Flutter application.
+        /// </remarks>
+        /// <param name="call">A <see cref="MethodCall"/>.</param>
+        /// <returns>A <see cref="Task"/> that completes with a result.</returns>
+        public async Task<object> InvokeMethodAsync(MethodCall call)
+        {
+            if (call == null)
+            {
+                throw new ArgumentNullException(nameof(call));
+            }
+            byte[] result = await BinaryMessenger.SendAsync(Name, Codec.EncodeMethodCall(call));
+            return Codec.DecodeEnvelope(result);
+        }
+
+        /// <summary>
+        /// Registers a method call handler on this channel.
+        /// </summary>
+        /// <param name="handler">A <see cref="MethodCallHandler"/>, or null to deregister.</param>
+        public void SetMethodCallHandler(MethodCallHandler handler)
+        {
+            async Task<byte[]> binaryHandler(byte[] bytes)
+            {
+                MethodCall call = Codec.DecodeMethodCall(bytes);
+                try
+                {
+                    return Codec.EncodeSuccessEnvelope(await handler(call));
+                }
+                catch (FlutterException fe)
+                {
+                    return Codec.EncodeErrorEnvelope(fe.Code, fe.Message, fe.Details, fe.StackTrace);
+                }
+                catch (Exception e)
+                {
+                    return Codec.EncodeErrorEnvelope("error", e.Message, null);
+                }
```

Comment:
Shouldn't this be like
```suggestion
                catch (FlutterException fe)
                {
                    return Codec.EncodeErrorEnvelope(fe.Code, fe.Message, fe.Details);
                }
                catch (Exception e)
                {
                    return Codec.EncodeErrorEnvelope("error", e.Message, null, e.StackTrace);
                }
```
Also if the user wants to indicate that a **certain** method is **NotImplemented**, should they set the method call handler of the **entire** channel to null? Is there something that corresponds to Dart's [MissingPluginException](https://api.flutter.dev/flutter/services/MissingPluginException-class.html) or Java's [MethodChannel.Result.notImplemented](https://api.flutter.dev/javadoc/io/flutter/plugin/common/MethodChannel.Result.html)?

Let's say there's a simple C++ plugin that only implements the `getPlatformVersion` method:

```cpp
channel->SetMethodCallHandler(
    [plugin = this](const auto &call, auto result) {
      plugin->HandleMethodCall(call, std::move(result));
    });

void HandleMethodCall(
    const flutter::MethodCall<flutter::EncodableValue> &method_call,
    std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> result) {
  if (method_call.method_name().compare("getPlatformVersion") == 0) {
    result->Success(flutter::EncodableValue("1.0"));
  } else {
    result->NotImplemented();
  }
}
```
Then how would you re-implement this plugin in C#?

Reply:
I missed the part of `NotImplemented()`. I will consider adding the `Result` interface.

Reply:
I added `MissingPluginException` instead of `Result` interface because Java style implementation is not appropriate when using the async function.

Reply:
#### example:
```c#
            channel.SetMethodCallHandler(async (call) =>
            {
                switch (call.Method)
                {
                    case "method1":
                        return new ArrayList() { 1, "TEST_STRING", true };
                    case "method2":
                        throw new FlutterException("my001", "test error!!", null);
                    default:
                        throw new MissingPluginException();
                }
            });
```

Reply:
Actually, we can use other system exceptions instead of adding new `MissingPluginException`. for example `System.MissingMethodException`,  `System.NotSupportedException` or `System.NotImplementedException`. @swift-kim how about your opinion?

Reply:
`MissingPluginException` feels slightly better for me for consistency with the Dart side implementation. (You could also replace `FlutterException` with `PlatformException` if you want.)

Reply:
> for consistency with the Dart side 
I agree to be consistent with Dart implementation. In this sense, `PlatformException` would be better. thanks.

Reply:
> I agree to be consistent with Dart implementation. ~In this sense, `PlatformException` would be better. thanks.~
Sorry, I would like to keep `FlutterException` again because it means that Platform and Flutter have opposite relationships.
* PlatformException : Thrown to indicate that a platform interaction failed in the platform plugin
* FlutterException: Thrown to indicate that a Flutter method invocation failed on the Flutter side.
In fact, `InvokeMethod` and `InvokeMethodAsync` of `MethodChannel` can throw this exception when an error response is received.

Reply:
Ah, I thought they were referring to the same thing. Thanks for the explanation.